Add phononic cavity tutorial

dangars opened the pull request:
Following up on the [google group conversation](https://groups.google.com/forum/#!topic/dealii/N-V-8geRLrA).

Todo list:

- [x] Use real/imag separation with UMFPack
- [x] Rewrite the equations in the complex domain and solve them with std::complex PETSc/MUMPS
- [x] std::complex PETSc/MUMPS gives the same result as real/imag separation with UMFPack 
- [x] Parallelize the code with MPI
- [x] Submit the hdf5 class in a different PR
- [x] Improve the documentation

dangars commented:
If the code is unclear, I can improve the documentation and re-submit the code.

jppelteret commented:
Hi @dangars. I just wanted to pop by and let you know that I've seen your PR. Thanks for being willing to contribute a new tutorial. I (and maybe some of the other devs) gather that since you have a todo list that you're still working on this. Please let us know when you're wanting us to discuss any points that you've listed above!

dangars commented:
Thanks @jppelteret  for your message! I made progress with the integration with complex/parallel petsc. I should be able to submit the code soon

These type of calculations can be useful for the optomechanics community. I know opensource code for the calculation of photonics structures (https://meep.readthedocs.io) but I don't know any opensource code for the calculation of phononics structures.

I'll let you know when the code is ready for discussion!

bangerth commented:
@dangars -- Yes, please do! Reviewing big patches always takes a lot of time, so people are reluctant to look at things that may not be quite finished yet. But when you think you have a state you'd like feedback on, let us know!

bangerth commented on examples/step-a/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+##
+
+# Set the name of the project and target:
+SET(TARGET "step-a")
```
you can take step-62 already!

dangars replied:
Done!

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2000 - 2016 by the deal.II authors
```
you'll want to just say 2018 here

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+
+ *
+ * Author: Wolfgang Bangerth, University of Heidelberg, 2000
+ *         Daniel Garcia-Sanchez, CNRS, 2018
```
and I don't know that I should get any credit for this either :-)

dangars replied:
Ok :)

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+    H5Group &operator=(H5Group &&other);
+    ~H5Group();
+    void
+                store_matrix(const std::vector<std::vector<std::complex<double>>> &matrix,
```
funny alignment here with the return type separated from the function name :-)

dangars replied:
I think that this is related to clang-format. The clang-format from the library is different than the format from the examples. I'll put this code in the library, therefore this should be fine. I'll make sure that the format is ok!

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+
+namespace hdf5
+{
+  class H5Group
```
I haven't even tried to understand yet what this class does, but do you think that that might be generic functionality better suited for the library proper, rather than an individual tutorial program?

dangars replied:
This class provides hdf5 functionality. Create groups, datasets, save attributes, read attributes, etc.
I use it to store the flux vs the frequency across a section and then read the data with python.
I'll submit the class in a separate PR, but first I'll improve the library, add support for deal.ii Vectors/Tensors and add MPI support.

bangerth replied:
Great, I think that would be useful if you could make it general enough to use in other contexts as well!

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+  H5Group &H5Group::operator=(const H5Group & /*other*/)
+  {
+    throw std::runtime_error(
+      std::string("Not implemented at Copy Assignment Operator H5Group"));
```
couldn't you just `=delete` this (and the next) operator?

bangerth replied:
oh, and the previous one as well? if they can't be called at run time, maybe we should just disallow them at compile time as well.

dangars replied:
Done!

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+    DoFHandler<dim> dof_handler;
+
+    // spacedim is 2*dim because of the real and imaginary parts
+    FESystem<dim> fe;
```
no, `spacedim==dim` -- it's just that the number of *components* is twice `dim` :-)

dangars replied:
Ok :)

bangerth commented on examples/step-a/step-a.cc:
```diff
@@ -0,0 +1,1654 @@
+            stiffness_tensor[i][j][k][l] =
+              (((i == k) && (j == l) ? mu : 0.0) +
+               ((i == l) && (j == k) ? mu : 0.0) +
+               ((i == j) && (k == l) ? lambda : 0.0));
```
This is not wrong, but I suspect that you can write this in simpler form using the predefined rank-4 tensors either in the physics module or the ones that are marked as `@related` to the `SymmetricTensor<4,dim>` type.

But it may be instructive to actually see this here. When you write the documentation, you can always refer to using the other functions.

dangars replied:
You are right! I think that I can use the double_contract operation from the Tensor class.
The tensor is non-symmetrical because of the PMLs.

dangars commented on examples/step-a/doc/intro.dox:
```diff
@@ -0,0 +1,92 @@
+<a name="Intro"></a>
+<h1>Introduction</h1>
+<h3>Micropillar superlattice cavity</h3>
+In this tutorial we are going to calculate the <a href="https://en.wikipedia.org/wiki/Band_gap">energy band gap</a> and the mechanical resonance of a <a href="https://doi.org/10.1103/PhysRevA.94.033813">micropillar superlattice cavity</a>.
```
Done!

dangars commented on examples/step-a/doc/intro.dox:
```diff
@@ -0,0 +1,92 @@
+The thickness of the mirror layers (brown and green) is $\lambda/4$ and the thickness of the cavity (blue) is $\lambda/2$.
+The micropillar cavity has a cylindrical shape as shown below:
+
+<img src="https://github.com/dangars/dealii/blob/phononic-cavity/examples/step-a/images/micropillar.jpg?raw=true">
```
Sounds good

dangars commented:
@bangert Thanks a lot for reviewing the code!
I made the requested modifications. I answered your comments, although they are marked as outdated by github.
I still have to improve the code and the documentation before a request another review. But I have a couple of questions. See my comments below

dangars commented:
How do I make sure that the step is compiled only if complex-petsc is available?
I didn't find the CMake variable DEAL_II_WITH_COMPLEX_PETSC
Should I add something like?
``` c++
#if !(defined(DEAL_II_WITH_COMPLEX_PETSC))
#  error DEAL_II_WITH_COMPLEX_PETSC required
#endif
```
And in CMakeLists.txt
``` cmake
IF(NOT (DEAL_II_WITH_COMPLEX_PETSC) OR NOT DEAL_II_WITH_P4EST OR DEAL_II_WITH_PETSC) # keep in one line
  MESSAGE(FATAL_ERROR "
Error! This tutorial requires a deal.II library that was configured with the following options:
    DEAL_II_WITH_PETSC = OFF
    DEAL_II_WITH_COMPLEX_PETSC = ON
    DEAL_II_WITH_P4EST = ON
However, the deal.II library found at ${DEAL_II_PATH} was configured with these options
    DEAL_II_WITH_PETSC = ${DEAL_II_WITH_PETSC}
    DEAL_II_WITH_COMPLEX_PETSC = ${DEAL_II_WITH_COMPLEX_PETSC}
    DEAL_II_WITH_P4EST = ${DEAL_II_WITH_P4EST}
which conflict with the requirements.
One or both of the aforementioned combinations of prerequisites are not met by your installation, but at least one is required for this tutorial step."
    )
ENDIF()
```

dangars commented:
Deal.ii with complex MUMPS/PETSc works very well. I obtained a considerable performance improvement!
The whole program with std::complex PETSc/MUMPS runs 3 times faster than the code with real/imag separation with MUMPS. I would like to use an iterative solver.

The system matrix is complex and non-symmetric. As you can see in the equation the matrix is not symmetric and the values can be anything in the complex field. The equation is very similar to the equation from step-8.

In the equation the following coefficients are complex.
- s[i] = 1+js''[i]
- xi = s[0]*s[1]*s[2]

xi and s[i] represent the PML absortion and it is zero outside the PML. In the PML the stress is not symmetric because the PML is not a physical material.

I tried to use an iterative solver, but I couldn't find a suitable PETSc preconditioner.
- Hypre does not work with complex PETSc
- BlockJacobi does not really converge
- Acording to the hypre documentation, hypre seems to support complex, but not with PETSc.
- Trilinos seems to have a good support for complex preconditioners with Ifpack2/Tpetra. Is Trilinos a good option for complex preconditioners for non-symmetric matrices?
- Deal.ii uses Ifpack. Is there any reason to use lpack instead of Ifpack2/Tpetra?

![linear_system](https://user-images.githubusercontent.com/25910642/42786429-923ae380-8956-11e8-93c8-1730d407dbaa.png)

dangars commented:
Sounds good! I'll submit another PR with DEAL_II_WITH_COMPLEX_PETSC

dangars commented:
Yes, PETSc + MUMPS is the easiest and works quite well!
BlockJacobi/GMRES works, but it takes a very large number of iterations.
I reviewed the literature of Electromagnetics. It seems quite challenging to prepare an efficient preconditioner for a frequency domain problem with PMLs. It looks like it is the same for this equation.

dangars commented:
I will investigate if there is a benefit to split things into real and imag parts and then use the diagonal blocks to build a good preconditioner.
Then we can decide if it is worthy to put it in this tutorial or make two tutorials and compare the results: complex vs real/imag separation.
That being said, I tested PETSc/MUMPs with this equation and large phononic structures and it works quite well.

The great benefit of using complex algebra is that it is possible to have a phase shift in the boundary conditions. I'm planning to use this feature with SLEPc to calculate the band structure

bangerth commented:
I think that for a tutorial program, it is good enough if you just choose one possible option for a solver -- preferably one that has not been shown in a previous tutorial program.

You always have the option of discussion how one may improve this in the "Possibilities for extensions" section, without making it a whole new tutorial program. You can, for example, take a look at how step-22 does this.

dangars commented:
That sounds good :)
I choose the std::complex algebra which is new and gives a compact set of equations
I'll the discuss the possibilities for extensions

Rombur commented:
@dangars If you add a jupyter notebook, would it possible to move it here https://github.com/dealii/dealii/tree/master/contrib/python-bindings/notebooks and add an entry in the index notebook.

dangars commented:
@Rombur How do you want to organize the notebooks? How about the tutorial-1.ipynb, tutorial-2.ipynb, etc are python only tutorials and the step-62.ipynb, step-63.ipynb are the step tutorials that have a jupyter notebook? That way people can refer to the C++ code of the step tutorial

How about the last commit ? I keep step-62.ipynb in the notebooks folder and in the steps folder. It is ok to have duplicate files in git. It does not double the space because they share the same hash.
[phononic-cavity branch/notebooks](https://github.com/dangars/dealii/tree/phononic-cavity/contrib/python-bindings/notebooks)
[phononic-cavity branch/step-62 folder](https://github.com/dangars/dealii/tree/phononic-cavity/examples/step-62)

Should we add the author and the copyright like in the step tutorials? If you add a blurb to your jupyter notebook, I will just do copy & paste in mine.

I think that it is better to have relative indices in index.ipynb that way there are no problems with branches and it works locally.
```
[step-62](step-62.ipynb)
```
instead of
```
[tutorial-1](https://github.com/dealii/dealii/tree/master/contrib/python-bindings/notebooks/tutorial-1.ipynb)
```

Rombur commented:
> @Rombur How do you want to organize the notebooks? How about the tutorial-1.ipynb, tutorial-2.ipynb, etc are python only tutorials and the step-62.ipynb, step-63.ipynb are the step tutorials that have a jupyter notebook? That way people can refer to the C++ code of the step tutorial

Good idea!

> How about the last commit ? I keep step-62.ipynb in the notebooks folder and in the steps folder. It is ok to have duplicate files in git. It does not double the space because they share the same hash.

What about we refer to the notebook in step-62 or we can create a soft link to the notebook. I am not worry about the space but in keeping every in sync. If we copy the files, there is a chance that they will diverge because we forget to change it at both place.

>Should we add the author and the copyright like in the step tutorials? If you add a blurb to your jupyter notebook, I will just do copy & paste in mine.

Will do

> I think that it is better to have relative indices in index.ipynb that way there are no problems with branches and it works locally.
Edit: I just realized that in github, if I right click on the jupyter link to open a new tab it does not work if it is a relative link. Let me know if you prefer that I use relative links or absolute links as you did.

Yeah that's the reason why I put the absolute link. I think most people interact will interact with the notebook on github first so we should make it work there.

dangars commented:
> What about we refer to the notebook in step-62 or we can create a soft link to the notebook. I am not worry about the space but in keeping every in sync. If we copy the files, there is a chance that they will diverge because we forget to change it at both place.

You can take a look to [phononic-cavity branch/notebooks](https://github.com/dangars/dealii/tree/phononic-cavity/contrib/python-bindings/notebooks)

In index.ipynb step-62 points to dealii/examples/step-62/step-62.ipynb

On the other hand, the linux/git symbolic link appears in GitHub as ../../../examples/step-62/step-62.ipynb. I think that a python programmer would understand that you have to follow the symbolic link.

I think that it is better to leave a symbolic link because people will surf and take a look to the notebooks that are inside the notebooks folder. Myself I didn't look at the index, I look directly at tutorial-1.ipynb

Rombur commented:
> On the other hand, the linux/git symbolic link appears in GitHub as ../../../examples/step-62/step-62.ipynb. I think that a python programmer would understand that you have to follow the symbolic link.

It's a pity that it doesn't show the file directly but we can live with that.

> I think that it is better to leave a symbolic link because people will surf and take a look to the notebooks that are inside the notebooks folder. Myself I didn't look at the index, I look directly at tutorial-1.ipynb

I totally agree that the symbolic link should be visible. The index is only useful if we start having more than 10 notebooks, then it becomes cumbersome to check every file.

bangerth commented:
I've had a phone call for an hour today with @dangars and he mentioned that this isn't ready in its current form. He'll let us know when it is!

In the meantime, thanks already for the tutorial, @dangars! We look forward to it being ready!

dangars commented:
@bangerth That is right! I made several improvements to the code:
* Use symmetry as boundary condition to simulate half of the domain
* Store the Mass and Stiffness matrices and the right hand side in PointMatrix which reduces considerably total simulation time because they can be reused for each frequency step.

I'm going to merge these and update the documentation. It should be ready very soon!

dangars commented:
@bangerth The tutorial is ready for review!

Sorry for the delay. I modified a lot the tutorial. I simplified the tutorial and improved the documentation.

You can use the following link that points to the compiled doxygen documentation. The tutorial has many figures.
[step_62 doxygen](http://htmlpreview.github.io/?https://github.com/dangars/dealii/blob/phononic-cavity-doc/compiled_doc/doxygen/deal.II/step_62.html)

dangars commented:
Do you think that the tutorial could make it to the 9.1 release? I just realized that the deadline is coming soon

bangerth commented:
Yes, I think we can make that.

Anyone willing to also take a stab at reviewing?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+    // `RightHandSide::data`, `PML::data` and `Parameters::data`. When a
+    // HDF5::Group is copied, it will point to the same HDF5 Group; this is
+    // achieved with the protected std::shared_ptr<hid_t>
+    // HDF5::Group::hdf5_reference.
```
I think you are referring to internals of the `HDF5::Group` class here, but from a user perspective this is not of interest. Can you rephrase this in a way that explains *what* is happening, not *why*?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+    // notebook, stored in `data` as HDF5 attributes and then read by the
+    // constructor.
+    const std::string        simulation_name;
+    bool                     save_vtu_files;
```
It is a bit confusing why some of the member variables of this class are `const` and some are not. Is there a good reason?

dangars replied:
Now all the variables are const.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+  // `PointHistory` class has already been used in step-18.
+
+  template <int dim>
+  class PointHistory
```
It's not clear to me why this class is called `PointHistory`. You're just caching data. What is the connection?

Also, I believe that what you store here are not the *assembled* matrices and right hand sides, but instead the data for a single cell. Is this right? Maybe you could put that into the documentation of the class?

dangars replied:
I agree PointHistory is not a good name. Now the class is named `QuadratureCache`. Is that a good name?

bangerth replied:
Yes!

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+  // This class returns the stiffness tensor of the material. For the sake of
+  // simplicity we consider the stiffness to be isotropic and homogeneous; only
+  // the density $\rho$ depends on the position. As we have previously done in
+  // step-8. The stiffness coefficients $c_{ijkl}$ can be expressed in function
```
This sentence is incomplete -- *what* have we already done previously?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+  // similar to the step-40 tutorial program. The main differences are:
+  // - The sweep over the frequency vector.
+  // - We save the stiffness and mass matrices in `quadrature_point_history` and
+  // use them for each frequency step.
```
The same question applies here to the naming. The term "history" seems inappropriate.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+  class ElasticWave
+  {
+  public:
+    ElasticWave(Parameters<dim> parameters_);
```
make this a `const` reference?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1468 @@
+
+  private:
+    void setup_system();
+    void assemble_system(double omega, bool calculate_quadrature_data);
```
here and below, you pass a lot of arguments by value. We have a convention that all input arguments are always marked as `const`. Can you make that change?

dangars commented on contrib/python-bindings/notebooks/index.ipynb:
```diff
@@ -8,7 +8,11 @@
+    "\n",
+    "## step-62\n",
+    "[step-62](https://github.com/dangars/dealii/tree/phononic-cavity/examples/step-62/step-62.ipynb) shows how to to calculate the [energy band gap](https://en.wikipedia.org/wiki/Band_gap) and the\n",
+    "mechanical resonance of a [micropillar superlattice cavity](https://doi.org/10.1103/PhysRevA.94.033813).\n"
```
I had a [conversation](https://github.com/dealii/dealii/pull/6747#issuecomment-411889703) with @Rombur about how to organize the notebooks. We agreed to to add an entry for each notebook in `index.ipynb`. In addition notebooks of the tutorials will be referenced by a symbolic link, see below. You can take a look to the conversation [conversation](https://github.com/dealii/dealii/pull/6747#issuecomment-411889703) and let us know if that is ok for you.

bangerth replied:
Ah, ok. I don't know anything about the ipnb files, and only thought that this was a mistake because none of the other programs are listed there. This works for me if you think it's right this way!

Rombur replied:
Before we merge this PR, we need to change the link to `https://github.com/dangars` to `https://github.com/dealii`

dangars replied:
@Rombur Thanks for seeing this. I updated the link from https://github.com/dangars to https://github.com/dealii I did a grep to make sure that there a no other links that point to dangars

dangars commented on contrib/python-bindings/notebooks/step-62.ipynb:
```diff
@@ -0,0 +1 @@
+../../../examples/step-62/step-62.ipynb
```
This is the symbolic link that I mentioned in [conversation](https://github.com/dealii/dealii/pull/6747#issuecomment-411889703).

dangars commented on examples/CMakeLists.txt:
```diff
@@ -31,6 +31,7 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
     PATTERN "*.cc"
     PATTERN "*.prm"
     PATTERN "*.inp"
+    PATTERN "*.ipynb"
```
The `*.ipynb` files are the jupyter notebooks. I would like to continue using jupyter notebooks for the tutorials that I will write. I find convenient to use jupyter/python notebooks for the data analysis.

dangars commented on examples/step-62/doc/intro.dox:
```diff
@@ -0,0 +1,220 @@
+In this tutorial we calculate the
+[energy gap](https://en.wikipedia.org/wiki/Band_gap) and the
+mechanical resonance of a
+[phononic superlattice cavity](https://doi.org/10.1103/PhysRevA.94.033813).
```
I didn't use FEM in that paper, I developed a semi-analytical method based on the scattering matrix for superlattice systems. As expected the finite element method with deal.ii gives the same results as the scattering matrix method that I developed.

dangars commented on examples/step-62/doc/intro.dox:
```diff
@@ -0,0 +1,220 @@
+href="../../readme.html" target="body">README</a> file.
+
+<h1>Introduction</h1>
+In this tutorial we calculate the
```
I agree! I improved the introduction :)

dangars commented on examples/step-62/doc/intro.dox:
```diff
@@ -0,0 +1,220 @@
+\nabla\cdot(\boldsymbol{\bar\sigma} \xi \boldsymbol{\Lambda})&=&-\omega^2\rho\xi\mathbf{\bar u}\\
+\boldsymbol{\bar \sigma} &=&\mathbf{C}\boldsymbol{\bar\varepsilon}\\
+\boldsymbol{\bar\varepsilon}&=&\frac{1}{2}[(\nabla\mathbf{\bar{u}}\boldsymbol{\Lambda}+\boldsymbol{\Lambda}^\mathrm{T}(\nabla\mathbf{\bar{u}})^\mathrm{T})]\\
+\xi &=&s_0\cdot s_1\cdot s_2\\
```
That is right. There is one coefficient per dimension. Now it is better explained in the documentation.

dangars commented on examples/step-62/doc/intro.dox:
```diff
@@ -0,0 +1,220 @@
+@f[
+\varepsilon_{kl} =\frac{1}{2}\left(\frac{1}{s_k}\partial_k u_l
++ \frac{1}{s_l}\partial_l u_k\right)
+@f]
```
The strain with the PML is not symmetric because of the `s_k` coefficients. The strain with the PML is slightly different that the strain from the standard elastic equation. Now I explain that in the text.

bangerth replied:
I see. I missed that, so others might as well. Maybe say something like this explicitly, rather than just showing a formula?

dangars commented on examples/step-62/doc/intro.dox:
```diff
@@ -0,0 +1,220 @@
+\varepsilon_{kl} =\frac{1}{2}\left(\frac{1}{s_k}\partial_k u_l
++ \frac{1}{s_l}\partial_l u_k\right)
+@f]
+which can be written as
```
The strain with the PML is not symmetric.

dangars commented:
@bangerth Thanks for the review :)
I addressed your comments.
I updated the link with the compiled version of the doxygen documentation:
[step_62 doxygen](http://htmlpreview.github.io/?https://github.com/dangars/dealii/blob/phononic-cavity-doc/compiled_doc/doxygen/deal.II/step_62.html)

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+
+
+  // @sect4{`Rho` class}
```
```suggestion
  // @sect4{Implementation of the `Rho` class}
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+  // This class is used to define the mass density. As we have explained, before
+  // a phononic superlattice cavity is formed by two
+  //[Distributed Reflector](https://en.wikipedia.org/wiki/Band_gap),
```
```suggestion
  // [Distributed Reflector](https://en.wikipedia.org/wiki/Band_gap),
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+  // @sect4{`Rho` class}
+
+  // This class is used to define the mass density. As we have explained, before
```
```suggestion
  // This class is used to define the mass density. As we have explained before,
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // a phononic superlattice cavity is formed by two
+  //[Distributed Reflector](https://en.wikipedia.org/wiki/Band_gap),
+  // mirrors and a $\lambda/2$ cavity where $\lambda$ is the acoustic
+  // wavelength. Acoustic DBRs are  periodic structures where a set of bilayer
```
```suggestion
  // wavelength. Acoustic DBRs are periodic structures where a set of bilayer
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // mirrors and a $\lambda/2$ cavity where $\lambda$ is the acoustic
+  // wavelength. Acoustic DBRs are  periodic structures where a set of bilayer
+  // stacks with contrasting physical properties (sound velocity index) is
+  // repeated $N$ times. The change of in the velocity will be obtained by
```
```suggestion
  // repeated $N$ times. The change in the wave velocity is generated by
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // @f]
+    // where $K_e$ is the effective elastic constant and $\rho$ the density.
+    // Here we consider the case in which the waveguide width is much smaller
+    // than the wavelength. In this case it can be shown that for a two
```
```suggestion
    // than the wavelength. In this case it can be shown that for the two
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // @f[
+    //  K_e = 4\mu\frac{\lambda +\mu}{\lamda+2\mu}
+    // @f]
+    // and for a three dimensional case $K_e$ is equal to the Young's modulus.
```
```suggestion
    // and for the three dimensional case $K_e$ is equal to the Young's modulus:
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+      material_b_speed_of_sound / cavity_resonance_frequency;
+
+    // The density $\rho$ takes the following form
+    //<img alt="Phononic superlattice cavity"
```
```suggestion
    // <img alt="Phononic superlattice cavity"
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // The density $\rho$ takes the following form
+    //<img alt="Phononic superlattice cavity"
+    // src="https://raw.githubusercontent.com/dangars/dealii/phononic-cavity/examples/step-62/doc/step-62.04.svg?sanitize=true"
+    // height="200" />
```
Before this program is committed, we should change these links to point to the deal.II image repository. If you send me all of the pictures you'd like to include in this program, then I can add them there and give you links to reference them!

dangars replied:
I just sent you an email :)
I'll update the links and remove the commit with the images, when they'll be in the deal.II image repository.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // represents material_b.
+    for (unsigned int idx = 0; idx < nb_mirror_pairs; idx++)
+      {
+        double layer_transition_center =
```
```suggestion
        const double layer_transition_center =
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+              (layer_transition_center - average_rho_width / 2) &&
+            std::abs(p[0]) <= (layer_transition_center + average_rho_width / 2))
+          {
+            double coefficient = (std::abs(p[0]) - (layer_transition_center -
```
```suggestion
            const double coefficient = (std::abs(p[0]) - (layer_transition_center -
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // which improves the precision of the simulation.
+    for (unsigned int idx = 0; idx < nb_mirror_pairs; idx++)
+      {
+        double layer_transition_center =
```
```suggestion
        const double layer_transition_center =
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+              (layer_transition_center - average_rho_width / 2) &&
+            std::abs(p[0]) <= (layer_transition_center + average_rho_width / 2))
+          {
+            double coefficient = (std::abs(p[0]) - (layer_transition_center -
```
```suggestion
            const double coefficient = (std::abs(p[0]) - (layer_transition_center -
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // the material_a layers
+    for (unsigned int idx = 0; idx < nb_mirror_pairs; idx++)
+      {
+        double layer_center =
```
same here and below -- everything that *could* be `const` *should* be `const`

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+
+
+  // @sect4{`Parameters` class}
```
```suggestion
  // @sect4{Implementation of the `Parameters` class}
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // @sect4{`Parameters` class}
+
+  // The constructor reads all the parameters from the HDF5::Group `data` using
+  // the HDF5::Group::get_attribute function.
```
```suggestion
  // the HDF5::Group::get_attribute() function.
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+
+
+  // @sect4{`QuadratureCache` class}
```
```suggestion
  // @sect4{Implementation of the `QuadratureCache` class}
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // We need to reserve enough space for the mass and stiffness matrices and the
+  // right hand side vector.
+  template <int dim>
+  QuadratureCache<dim>::QuadratureCache(unsigned int dofs_per_cell)
```
```suggestion
  QuadratureCache<dim>::QuadratureCache(const unsigned int dofs_per_cell)
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  ElasticWave<dim>::~ElasticWave()
+  {
+    dof_handler.clear();
+  }
```
Do you actually need this? Usually, if you just declare the finite element object *before* the DoFHandler, then you don't need the call to `clear()`, and consequently don't need a destructor at all.

dangars replied:
I copied this code from step-40. I can add a PR to modify step-40 accordingly :)

I think that it would be good to document in DoFHandler.clear() that there is no need to call this function in the destructor if the finite element object has been declared before the dof_handler. I've not see it in the documentation.

bangerth replied:
Sure, if you feel like modifying step-40 in a separate PR, we'd take that as well :-)

Thinking about it, it doesn't actually depend on the order of the objects any more. That used to be the case, but it's no longer the case. This should simply *always* work without the call to `clear()`.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+  // @sect4{ElasticWave::assemble_system}
+
+  // This very similar to step-40. Although there are notable differences. We
```
```suggestion
  // This function is also similar to step-40, though there are notable differences. We
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // @sect4{ElasticWave::assemble_system}
+
+  // This very similar to step-40. Although there are notable differences. We
+  // assembly the system for each frequency/omega step. In the first step we set
```
```suggestion
  // assemble the system for each frequency/omega step. In the first step, the second argument of the function is
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+            Assert(local_quadrature_points_data >= &quadrature_cache.front(),
+                   ExcInternalError());
+            Assert(local_quadrature_points_data < &quadrature_cache.back(),
+                   ExcInternalError());
```
Shouldn't the comparison be `<=`? I'm uncertain how this works...

dangars replied:
I'm not sure either how it works. I copied the code from step-18. I replaced `>=` by `<=` and the program complains in debug mode, thus `>=` seems to be correct.

bangerth replied:
The `>= &quadrature_cache.front()` seems correct to me, but since `back()` is an actual element (and not the past-the-end element), I would have expected that the correct second condition is `<= &quadrature_cache.back()` rather than `< &quadrature_cache.back()`. What happens if you make this change?

bangerth replied:
Ah, I understand now: We resize this array to
```
  quadrature_cache.resize(our_cells * quadrature_formula.size(),
                                           QuadratureCache<dim>(fe.dofs_per_cell));
```
so for every cell there are `quadrature_formula.size()` elements. We get a pointer to the first of these elements above, and so as long as `quadrature_formula.size() > 1`, the pointer we obtain is never a pointer to the last element of the array. So the assertion succeeds, but it should fail if one had a 1-point quadrature_formula. Indeed, I've tried this out and that's what happens. I'll fix step-18. Can you make the same fix here?

bangerth replied:
See #7960.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+  // We use this function to calculate the values of the position vector.
+  template <int dim>
+  void ElasticWave<dim>::set_position_vector()
```
I don't recall whether I commented on this before or not: a function that starts with `set_` generally (by convention, not necessity) takes an argument and sets a member variable. I think that you use this function *initialize* something, right? Maybe call it `initialize_position_vector()`.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+        // Because of the way the operator + and - are overloaded. To substract
+        // two points, the following has to be done:
+        // `Point_b<dim> + (-Point_a<dim>)`
+        p = (position_idx / ((double)(parameters.nb_probe_points - 1))) *
```
move the declaration of `p` here. Then you can also make it `const`.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+              (parameters.probe_stop_point + (-parameters.probe_start_point)) +
+            parameters.probe_start_point;
+        position[position_idx][0] = p[0];
+        position[position_idx][1] = p[1];
```
If you made `position` a `std::vector<Point<dim>>` as I commented on above, then this would simply be a straight assignment -- or, even better, you don't need the `p` variable at all :-)

dangars replied:
I'll do it after the support for `Point<dim>` in HDF5 has been merged :)

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+    // We store the displacement in the $x$ direction; the displacement in the
+    // $y$ direction is negligible.
+    const int probe_displacement_component = 0;
```
```suggestion
    const unsigned int probe_displacement_component = 0;
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+        Point<dim> point;
+        for (unsigned int dim_idx = 0; dim_idx < dim; ++dim_idx)
+          {
+            point(dim_idx) = position[position_idx][dim_idx];
```
```suggestion
            point[dim_idx] = position[position_idx][dim_idx];
```
(For consistency with the other places where you index into a `Point`.)

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+    // `std::complex<double>`.
+    else
+      {
+        displacement.write_none<std::complex<double>>();
```
nice illustration of `write_none()`! You might want to consider linking from that function to this program.

dangars replied:
I'll do that :)

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+        displacement.write_none<std::complex<double>>();
+      }
+
+    // If the variable of the jupyter notbook `save_vtu_files = True` then all
```
Let's drop the reference to how the parameter file was created
```suggestion
    // If the variable `save_vtu_files` in the input file equals `True` then all
```

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+                for (unsigned int dim_idx = 0; dim_idx < dim; ++dim_idx)
+                  {
+                    force[dim_idx](cell->active_cell_index()) =
+                      parameters.right_hand_side.value(cell->center(), dim_idx);
```
why not set this to `-1e20` as well?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+        unsigned int      nb_number_positions;
+        std::stringstream frequency_idx_stream;
+        nb_number_positions =
+          ((unsigned int)std::log10(parameters.nb_frequency_points)) + 1;
```
just move the assignment into the initialization of the variable. then it can also be `const`

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+    {
+      std::vector<QuadratureCache<dim>> tmp;
+      tmp.swap(quadrature_cache);
```
I had to do a double-take on this. I think it would be easier to read as
```suggestion
  quadrature_cache.swap (tmp);
```
If you feel like it, feel free to post a similar patch for step-18 :-)

dangars replied:
I'll submit a PR!

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+         cell != triangulation.end();
+         ++cell)
+      if (cell->is_locally_owned())
+        ++our_cells;
```
You can get that number via `triangulation.n_locally_owned_cells()` :-)

dangars replied:
I didn't find that function in [Triangulation](https://www.dealii.org/current/doxygen/deal.II/classTriangulation.html). I can submit a PR with that function, if you find that function, if you think that it is useful.

bangerth replied:
It's in `parallel::TriaBase` I believe. You need a parallel triangulation of some kind for this, but you have one.

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+  // @sect4{ElasticWave::frequency_sweep}
+  template <int dim>
+
+  // For clarity we divide the function `run` of step-40 into the functions
```
Move the lonely `template <int dim>` after the comments :-)

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+      {
+        std::cout << parameters.simulation_name + " frequency idx: "
+                  << frequency_idx << '/' << parameters.nb_frequency_points - 1
+                  << std::endl;
```
Don't you have to use `pcout` here and below?

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+          }
+
+        // We calculate the frequency and omega values for this particular step.
+        double current_loop_frequency =
```
make it `const`. same for the next variable

bangerth commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+    setup_quadrature_cache();
+
+    set_position_vector();
```
I've commented above that maybe a better name for this function would be `initialize_position_vector()`. An alternative, that matches the previous call here, would be `setup_position_vector()`.

bangerth commented on examples/step-62/doc/results.dox:
```diff
@@ -0,0 +1,67 @@
+h5_file.close()
+@endcode
+
+As expected the phononic superlattice cavity exhibits a mechanical resonance at 20GHz and a quality factor of 5046.
```
```suggestion
Given the values we have chosen for the parameters, one could estimate resonance frequencies
and quality factors analytically. Indeed, these are then confirmed by what we get in this program:
the phononic superlattice cavity exhibits a mechanical resonance at 20GHz and a quality factor of 5046.
The following images show the transmission amplitude and phase as a function of frequency in the
vicinity of the resonance frequency:
```

bangerth commented on examples/step-62/doc/results.dox:
```diff
@@ -0,0 +1,67 @@
+<img alt="Phononic superlattice cavity" src="https://raw.githubusercontent.com/dangars/dealii/phononic-cavity/examples/step-62/doc/step-62.05.svg?sanitize=true" height="400" />
+<img alt="Phononic superlattice cavity" src="https://raw.githubusercontent.com/dangars/dealii/phononic-cavity/examples/step-62/doc/step-62.06.svg?sanitize=true" height="400" />
+
+To obtain the phononic bandgap around the mechanical resonance, the parameter frequency range can be set to 16 GHz.
```
```suggestion
The images above suggest that the periodic structure has its intended effect: It really only lets waves of a very
specific frequency pass through, whereas all other waves are reflected. This is of course precisely what one builds
these sorts of devices for.

But it is not quite this easy. In practice, there is really only a "band gap", i.e., the device blocks waves other than
the desired one at 20GHz only within a certain frequency range. Indeed, to find out how large this "gap" is within
which waves are blocked, we can extend the frequency range to 16 GHz through the appropriate parameters in the
input file. We then obtain the following image:
```

bangerth commented on examples/step-62/doc/results.dox:
```diff
@@ -0,0 +1,67 @@
+
+To obtain the phononic bandgap around the mechanical resonance, the parameter frequency range can be set to 16 GHz.
+
+<img alt="Phononic superlattice cavity" src="https://raw.githubusercontent.com/dangars/dealii/phononic-cavity/examples/step-62/doc/step-62.07.svg?sanitize=true" height="400" />
```
```suggestion
<img alt="Phononic superlattice cavity" src="https://raw.githubusercontent.com/dangars/dealii/phononic-cavity/examples/step-62/doc/step-62.07.svg?sanitize=true" height="400" />

What this image suggests is that in the range of around 18 to around 22 MHz, really only the waves with a frequency
of 20 MHz are allowed to pass through, but beyond this range, there are plenty of other frequencies that can pass
through the device. If one wanted to filter those out as well, one would need to combine the device we have modeled
here with others that block the remainder of the unwanted waves that may be part of the input signal.
```

dangars replied:
I like your suggestions for the results section!
I expanded the discussion in the results section

dangars commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+    // This vector contains the coordinates $(x,y)$ of the points of the
+    // measurement probe.
+    FullMatrix<double> position;
```
I renamed it to probe_positions.

I do agree that `std::vector<Point<dim>>` makes more sense!
Before I use `std::vector<Point<dim>>` I have to add support in HDF5 for that container. I save the variable `probe_positions` in the HDF5 file. How about we leave it for now like this until I add the support in HDF5 for `std::vector<Point<dim>>`. Later I can modify this in another PR.

I think that the best way to save `std::vector<Point<dim>>` in HDF5 is to consider Point<dim> as a compound HDF5 type. The advantages are:
* `std::complex<double>` is already is already a compound type
* All the machinery is already in place. I just have to add the types `Point<1>`, `Point<2>` and `Point<3>` to HDF5::internal::get_hdf5_datatype()
* It will be possible to create 1D, 2D or 3D datasets of Points. In this tutorial I used a 1D dataset of Points. In another I may want to have a 2D probe, in that case I may need a 2D dataset of Points.
* Python understands quite well HDF5 compound types

bangerth replied:
I think that's an ok approach if you're willing to address this in the near future. If you want, you can always open github issues to remind yourself of these things.

dangars commented on examples/step-62/step-62.cc:
```diff
@@ -0,0 +1,1467 @@
+
+
+
+  // @sect4{`PML` class}
```
Can you take a look to the the general table of context with the link below and let me know if it looks ok?
`@sect4{PML class}` is alread inside `@sect3{Implementation of the auxiliary classes and functions} `

[step_62 doxygen](http://htmlpreview.github.io/?https://github.com/dangars/dealii/blob/phononic-cavity-doc/compiled_doc/doxygen/deal.II/step_62.html)

bangerth replied:
Ah, fair point. Then maybe just `"The PML class"`?

dangars commented:
@bangerth Thanks a lot for your review :)
I addressed your comments and expanded the results section. I'll remove the commit with the images when you upload the images to the deal.ii server.

I updated the link with the compiled version of the doxygen documentation:
[step_62 doxygen](http://htmlpreview.github.io/?https://github.com/dangars/dealii/blob/phononic-cavity-doc/compiled_doc/doxygen/deal.II/step_62.html)

dangars commented:
@bangerth I removed the images, updated the image links and addressed your comments.

The last compiled version is at:
[step_62 doxygen](http://htmlpreview.github.io/?https://github.com/dangars/dealii/blob/phononic-cavity-doc/compiled_doc/doxygen/deal.II/step_62.html)

bangerth commented:
Let's do this. You have a bullet point list with unchecked items at the very top of this post. Let us know if you want us to not merge this PR because of them, though I feel like everything that should be addressed in this PR has indeed already been addressed.

/rebuild

dangars commented:
@bangerth I updated the bullet points. This PR can be merged. I agree we addressed everything that should be addressed :)

I didn't addressed the following bullet points, but that is ok. It can be merge without these bullet points.
- **Use an iterative solver and a preconditioner** As we discussed, let's keep it simple and use MUMPS. I believe that `std::complex<double>` MUMPS has not been used in another tutorial. I'll do some research and check if the preconditioner "shifted Laplacian" that I found in the literature works. If it works I will added as "Possible extension" or in another tutorial.
- **Improve the efficiency of the probe** I think that the way I take the probe is not very efficient because at every frequency step I call `VectorTools::point_value`. This function loops over all the cells. I think that I can create a "cache" that maps the "probe_points" to the cell numbers.  I have to take some time to think about the best way to do it and to do some tests. I can eventually add an issue. On the other hand I think that the amount of time spend in that part of the code is negligible.

bangerth commented:
Sounds good -- let's merge!

Your plans for extensions sounds good. For the efficiency of the point evaluation, take a look at how I solved this in the new code gallery program MCMC-Laplace.

bangerth commented:
Once this is merged, please do a follow-up PR that adds an entry into `doc/news/changes/major`.

dangars commented on examples/step-62/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+    DEAL_II_WITH_PETSC = ON
+    DEAL_II_PETSC_WITH_COMPLEX = ON
+    DEAL_II_WITH_P4EST = ON
+    DEAL_II_WITH_HDF5 = ON
```
I added these dependencies. This is based on the CMakeLists.txt of step-40.

bangerth commented:
Ah, good point. Let's see what the tester has to say before we merge.

bangerth commented:
OK, it's done! Another tutorial program is in!

dangars commented:
Great!

@bangerth Thanks a lot for the review, it improved a lot the quality of the tutorial :)

jppelteret commented:
I'm glad to see that this new tutorial is done! Awesome contribution @dangars, thanks a lot!

I've noticed that `step-62` doesn't appear on the summary page for the tutorials (https://dealii.org/developer/doxygen/deal.II/Tutorial.html). I believe that the correct file to add this to is `doc/doxygen/tutorial/tutorial.h.in`.

dangars commented:
@jppelteret Thanks for your message!
I'm happy to see that the tutorial is useful :)
I'm going to submit another PR for the update of `doc/doxygen/tutorial/tutorial.h.in`
